### PR TITLE
Fix shadows for billboarded Sprite3D's

### LINKED ISSUE #41420: Shadows for Sprite3D with Billboard enabled always face the Vector3(0, 0, 0) rotational degree angle
I saw other issues like https://github.com/godotengine/godot/issues/27738 that is similar, but it's not exact the same.

**Godot version:**

3.2.2

**OS/device including version:**

Windows 10. GLES3

**Issue description:**

Hello,
When I have Sprite3D objects with Billboard enabled, the shadow is not facing the correct angle based on the DirectionalLight. It always faces the default Vector3(0, 0, 0) rotational degree angle of the Sprite3D, even if I rotate the object.
The image below has 8 Sprite3D.
Row 1 objects have Billboard enabled. Row 2 objects have Billboard disabled.
Column 1 have a rotational vector3 at (0, 0, 0) on both rows.
Column 2 have (0, 90, 0).
Column 3 have (0, 180, 0).
Column 4 have (0, 270, 0).
I would expect the shadow at the very least, to be drawn based on the objects rotational degree if Billboarding isn't supposed to work. I'm not very familiar with billboarding, but should the shadows behave like this?

image

**Minimal reproduction project:**

[Billboard Shadow Test.zip](https://github.com/godotengine/godot/files/5107210/Billboard.Shadow.Test.zip)

### LINKED ISSUE #69026: Weird shadows on Quad/Plane/Sprite3D with billboarding enabled
### Godot version

4.0.beta5.official

### System information

Pop!_OS 22.04 LTS x86_64, NVIDIA GeForce GTX 1060 6GB, Vulkan

### Issue description

These weird shadows appear across Sprite3D and quad & plane MeshInstances when they have a material with billboarding enabled.

kuva

### Steps to reproduce

1. Add a quad/plane MeshInstance node or a Sprite3D node to the scene
2. Add a new StandardMaterial3D or ORMMaterial3D to the mesh or material override. If using Sprite3D, also add a texture to the texture field, such as icon.svg.
3. Enable 'Billboard' on the material

### Minimal reproduction project

[4.0 Billboard Bug.zip](https://github.com/godotengine/godot/files/10070905/4.0.Billboard.Bug.zip)

### PR DESCRIPTION
**Pull request by ecmjohnson**
This resolves #69026 and resolves #41420 in the `master` branch based on my reading of @clayjohn's [suggested approach](https://github.com/godotengine/godot/issues/41420#issuecomment-834717315): the "main" camera transform is piped into the shader and the billboard always faces that transform, even when rendering shadow passes. I tested with both forward and mobile rendering backends.

fixed_billboard_sprite_shadows_small

The `main_cam_inv_view_matrix` needed is added to [`struct SceneData`](https://github.com/godotengine/godot/blob/a16b0fec4093b22ef5f14c1a4eb9363c890ab06d/servers/rendering/renderer_rd/shaders/scene_data_inc.glsl#L6) since it seemed like "scene data" and there might be a future case for using it elsewhere in the shader. If it would make more sense for a conditional `uniform` to be used (like for other material features), I can change that.

### CODE REVIEW

**Comment by ecmjohnson** [comment]
Sorry, my mistake; I see [the page on code style now](https://docs.godotengine.org/en/stable/community/contributing/code_style_guidelines.html) and I'll set up my format checker

File: servers/rendering/renderer_rd/storage_rd/render_scene_data_rd.cpp [hunk 1]
```diff
@@ -76,6 +76,8 @@ void RenderSceneDataRD::update_ubo(RID p_uniform_buffer, RS::ViewportDebugDraw p
 		ubo.eye_offset[v][3] = 0.0;
 	}
 
+	RendererRD::MaterialStorage::store_transform(main_cam_transform, ubo.main_cam_inv_view_matrix);
```
Comment (ecmjohnson) [inline servers/rendering/renderer_rd/storage_rd/render_scene_data_rd.cpp:79]: I'm matching the [above line](https://github.com/godotengine/godot/blob/master/servers/rendering/renderer_rd/storage_rd/render_scene_data_rd.cpp#L59) for storing the inverse of a `Transform3D` without calling `.affine_inverse()`, but I am a bit confused about that

Comment (ecmjohnson) [reply servers/rendering/renderer_rd/storage_rd/render_scene_data_rd.cpp:79]: Ah, I figured it out: in the 3.x branch there is only the camera matrix which is the inverse of the view matrix here. The view matrix is world to camera transformation, whereas the camera matrix is camera to world and they are inverses of each other

**Review by clayjohn — COMMENTED** [review]
This looks great! 

I am unsure about the naming of the new builtin so I would like to discuss this with other rendering contributors as soon as we are back to merging new features/ enhancements. Accordingly I am bumping this to the 4.1 milestone

**Comment by ecmjohnson** [comment]
@clayjohn I do think this is more of a bug than an enhancement since the shadows will be incorrect for billboarded Sprite3D's without this change. Currently the only way to accomplish billboarded Sprite3D's with shadows is by employing a [workaround](https://github.com/godotengine/godot/issues/27738#issuecomment-841782318)

**Comment by ZainWD** [comment]
Hope this gets implemented soon...

File: scene/resources/material.cpp [hunk 2]
```diff
@@ -988,18 +988,20 @@ void BaseMaterial3D::_update_shader() {
 		case BILLBOARD_DISABLED: {
 		} break;
 		case BILLBOARD_ENABLED: {
-			code += "	MODELVIEW_MATRIX = VIEW_MATRIX * mat4(INV_VIEW_MATRIX[0], INV_VIEW_MATRIX[1], INV_VIEW_MATRIX[2], MODEL_MATRIX[3]);\n";
+			// MAIN_CAM_INV_VIEW_MATRIX is inverse of the camera, even when rendering shadows: this ensures the billboard faces the camera when casting shadows
```
Comment (AThousandShips) [inline scene/resources/material.cpp:991]:
```suggestion
			// MAIN_CAM_INV_VIEW_MATRIX is inverse of the camera, even when rendering shadows: this ensures the billboard faces the camera when casting shadows.
```
As per the [comment style](https://docs.godotengine.org/en/latest/contributing/development/code_style_guidelines.html#comment-style-guide), same elsewhere

**Comment by AThousandShips** [comment]
@ZainWD Please don't bump PRs without contributing significant new information. Use the :+1: reaction button on the first post instead.

**Comment by BastiaanOlij** [comment]
I don't think this is a good solution. If you're making a 2 player game with side by side screen (yeah people still make those), or if you have additional viewports for other reasons, you'll have two different camera that both show the billboard in a different way, but only for one shadows will be correct.

Billboard is a technique that doesn't work well with shadows. If you're using it for LOD purposes, a billboard is only used when the object is far enough away that a shadow doesn't matter.

If you're using it for something like a name above a player or a healthbar, than it just can't cast a shadow. In this case the object should not be billboarded in the shader but you should orient it in code.

**Comment by ecmjohnson** [comment]
@BastiaanOlij You're totally correct that this wouldn't make sense in any scenario with multiple cameras in the scene. I'd add to your example stereo rendering, where the shadows would be generated based on billboarding to the e.g. left camera and be slightly incorrect for the other eye. I also agree with you that most uses of billboards should not use shadows and will not benefit from this change, but this use case can simple disable shadows.

However, I disagree that this makes this solution not worth having. As is, a beginner trying to create a single camera sprite game with 3D lighting (a la [Octopath Traveler](https://en.wikipedia.org/wiki/Octopath_Traveler) or [Cult of the Lamb](https://en.wikipedia.org/wiki/Cult_of_the_Lamb)) will have to rotate the sprites in software _themselves_. I think this use case should work as intuitively out-of-the-box to avoid hurting a beginner who simply wants their `Sprite3D`s to cast shadows. More advanced users making a two-player or stereo game should have the wherewithal to identify the problem when their shadows are weird from one perspective.

**Comment by Calinou** [comment]
I think this change makes sense, but the class reference must be modified to mention that this approach won't work in multi-viewport or XR scenarios.

File: scene/resources/material.cpp (hunk 2 above)
Comment (ecmjohnson) [reply scene/resources/material.cpp]: Corrected comment style

**Comment by ecmjohnson** [comment]
@Calinou That would be a separate PR in [the docs repo](https://github.com/godotengine/godot-docs) eh? I'll look into updating the `Sprite3D` documentation to clarify this behaviour

**Comment by BastiaanOlij** [comment]
@ecmjohnson for stereo I'm not worried, we're using multiview so the billboard will be aimed between the eye and the shadow will be based on that position. The problem is that in stereo you can't use billboards at all because our brain is waaaaay too smart. It immediately picks up on the billboard being used. It's hilarious actually, if you have a forest of trees in the background all rendered with billboard, this is PERFECT on screen, but in VR you see the trees rotated with your head.

Depth perception is trippy man :)

No I'm talking about multiplayer games. That said, I was already corrected by Clay who reminded me that we actually keep shadow maps per viewport. While in certain scenarios thats pretty wasteful, in others its exactly what we need. That is a topic for an entirely different day.

Still that said, I still believe this will bite us in the arse long term. Billboards and shadow just don't mix. It's way better to actually position a quad properly in your scene if this is the effect you're after instead of using visual trickery. Though I might be to purist and not pragmatical enough there :)

**Comment by BastiaanOlij** [comment]
In VR we always use quads that are properly positioned and aren't view aligned so there is stability in orientation regardless of the player turning their heads.

**Comment by Calinou** [comment]
> @Calinou That would be a separate PR in [the docs repo](https://github.com/godotengine/godot-docs) eh? I'll look into updating the `Sprite3D` documentation to clarify this behaviour

Class reference changes should be done in `doc/Classes/Sprite3D.xml`, not in godot-docs.

**Comment by ecmjohnson** [comment]
Sorry for the delay in such a simple change: it's currently summer in Canada. I [changed `docs/classes/BaseMaterial3D.xml`](https://github.com/ecmjohnson/godot/blob/709b4cb58b506964f1409c219996e7c4d64dc59c/doc/classes/BaseMaterial3D.xml#L128) to include an additional note reading:
```
			[b]Note:[/b] When billboarding is enabled and the material also casts shadows, billboards will face [b]the[/b] camera in the scene when rendering shadows. In scenes with multiple cameras, the intended shadow cannot be determined and this will result in undefined behaviour. See [url=https://github.com/godotengine/godot/pull/72638]GitHub Pull Request #72638[/url] for details.
```
I linked this PR since that could help a developer if they wanted to determine for which camera the shadow would be rendered (e.g. always wanting the shadow to be from a specific camera even with multiple cameras in the scene), but in most cases I would expect the billboards enabled and shadows enabled combination of settings to be a mistake when multiple cameras are used (e.g. couch co-op).

**Comment by ecmjohnson** [comment]
I also added the same note under [`docs/classes/SpriteBase3D.xml`](https://github.com/godotengine/godot/compare/master...ecmjohnson:godot:billboard_sprite_shadows#diff-6778f61dee233f69c95d8b99108b0496d415f1ec2b5bd441e314eb6bade63213R61) since that seems to be the one that shows in the editor when modifying the `Sprite3D` instance

**Comment by ecmjohnson** [comment]
It might be worth refactoring `RenderForwardClustered::_render_shadow_pass()` and `RenderForwardClustered::_render_shadow_append()` to take a `struct` parameter since the number of individual parameters is getting pretty nasty and this is a common point of extensibility for new features or bug fixes. I know this PR makes it one worse :flushed: but I've noticed because I've had to manually rebase a couple times now

**Comment by ecmjohnson** [comment]
I also missed making this fix to the compatibility rendering backend so I've added those now and verified the 3D billboarded sprites shadows on all rendering backends

**Review by clayjohn — APPROVED** [review]
Sorry for taking such a long time to review this. It took awhile to discuss with other contributors and then I kept poking at it and dropping it again. 

Overall, I think the implementation looks great. I don't think there is a better way to implement this and I think you have done everything correctly.

**Review by AThousandShips — APPROVED** [review]

**Comment by akien-mga** [comment]
While fixing the couple spelling issues in the docs, I would suggest rebasing on latest `master` to make sure it still builds fine after the last few weeks of active development on the rendering architecture.

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**